visual_behavior/dimensionality_reduction/clustering/figures.py

```diff
@@ -0,0 +1,298 @@
+for cre_line in cre_lines:
+    plotting.plot_coclustering_matrix_sorted_by_cluster_size(coclustering_matrices, cluster_meta, cre_line,
+                                                             save_dir=base_dir, folder=folder, ax=None)
+
+# plot average dropout scores for each cluster
+
+# plot each cluster separately and save to single cell examples dir
+cell_examples_dir = os.path.join(save_dir, 'matched_cell_examples')
+if not os.path.exists(cell_examples_dir):
+    os.mkdir(cell_examples_dir)
+
+# plotting.plot_dropout_heatmaps_and_save_to_cell_examples_folders(cluster_meta, feature_matrix, save_dir)
+
+# plot average dropouts for each cre line in cluster size order
+sort_col = 'cluster_id'
+plotting.plot_dropout_heatmaps_for_clusters(cluster_meta, feature_matrix, sort_col=sort_col, save_dir=base_dir,
+                                            folder=folder)
+
+# # plot dropout heatmaps in manually sorted order
+# sort_col = 'manual_sort_order'
+# plotting.plot_dropout_heatmaps_for_clusters(cluster_meta, feature_matrix, sort_col=sort_col, save_dir=base_dir,
+#                                             folder=folder)
+
+# plot feature matrix sorted by cluster ID
+plotting.plot_feature_matrix_sorted(feature_matrix, cluster_meta, sort_col='cluster_id', save_dir=base_dir,
+                                    folder=folder)
+
+# plotting.plot_feature_matrix_sorted(feature_matrix, cluster_meta, sort_col='manual_sort_order', save_dir=base_dir,
+#                                     folder=folder)
+
+# umap with cluster labels
+label_col = 'cluster_id'
+
+plotting.plot_umap_for_clusters(cluster_meta, feature_matrix, label_col=label_col, save_dir=base_dir, folder=folder)
+
+# Correlations within clusters
+plotting.plot_within_cluster_correlations_all_cre(cluster_meta, n_clusters_cre, sort_order=None, save_dir=base_dir,
+                                                  folder=folder)
+
+# average dropouts per cre line
+plotting.plot_average_dropout_heatmap_for_cre_lines(results_pivoted, cluster_meta, save_dir=base_dir, folder=folder)
+
+# # plot 100 cells per cluster to examine within cluster variability
+# dropouts = results_pivoted.merge(cells_table[['cell_specimen_id', 'experience_level', 'ophys_experiment_id']], on=['cell_specimen_id', 'experience_level'])
+# dropouts = dropouts.drop_duplicates(subset=['cell_specimen_id', 'ophys_experiment_id'])
+# plotting.plot_random_subset_of_cells_per_cluster(cluster_meta, dropouts, save_dir)
+
+
+# breakdown by area and depth
+
+# We are going to normalize within each area or depth to account for the large imbalance in N due to Scientifica datasets only being performed in V1 at certain depths, as well as biological variation in cell type specific expression by depth
+
+# plot fraction and number of cells by area and depth
+label = 'fraction of cells'
+plotting.plot_fraction_cells_by_area_depth(cluster_meta, n_clusters_cre, normalize=True, label=label,
```

Comment:
@matchings just a note that I have been splitting dis by cre line before providing them as an input to functions. Your functions expect 'cre_line' column. This can cause potential confusion.

Reply:
good point, we should reformat this, along with consolidating across all the "load_GLM_output" type functions that Alex identified as redundant. but i think we should do it in a different PR that is focused on finalizing the figure 4 functions, since this one is already huge.